File: auth-sdk/src/token-manager.ts
import { AxiosInstance } from "axios";

export interface StorageAdapter {
  getItem: (key: string) => Promise<string | null>;
  setItem: (key: string, value: string) => Promise<void>;
  removeItem: (key: string) => Promise<void>;
}

export class TokenManager {
  private currentToken: string | null = null;
  private currentTokenExpiry = 0;
  private tokenRequestPromise: Promise<string | null> | null = null;
  public onInvalidSession?: () => void;

  constructor(
    private api: AxiosInstance,
    private platform: string,
    private storage?: StorageAdapter,
  ) {}

  setCurrentToken(token: string | null) {
    this.currentToken = token;
    this.currentTokenExpiry = Date.now() + 15 * 60 * 1000;
  }

  async fetchToken(): Promise<string | null> {
    try {
      const headers: Record<string, string> = {
        "x-platform": this.platform,
      };

      if (this.platform !== "web" && this.storage) {
        const sessionId = await this.storage.getItem("sessionId");
        if (sessionId) {
          headers["x-session-id"] = sessionId;
        }
      }

      const response = await this.api.post("/token", null, {
        headers,
        withCredentials: true,
      });

      if (response.data.success) {
        if (
          this.platform !== "web" &&
          this.storage &&
          response.data.sessionId
        ) {
          await this.storage.setItem(
            "sessionId",
            String(response.data.sessionId),
          );
        }
        return response.data.token;
      }
      return null;
    } catch (error) {
      console.error("Error fetching token:", error);
      throw error;
    }
  }

  async getAuthToken(): Promise<string | null> {
    if (this.currentToken && Date.now() < this.currentTokenExpiry - 10000) {
      return this.currentToken;
    }

    if (this.tokenRequestPromise) {
      return this.tokenRequestPromise;
    }

    this.tokenRequestPromise = (async () => {
      try {
        const token = await this.fetchToken();
        if (token) {
          this.currentToken = token;
          this.currentTokenExpiry = Date.now() + 15 * 60 * 1000;
        } else {
          this.currentToken = null;
          this.currentTokenExpiry = 0;
          if (this.onInvalidSession) this.onInvalidSession();
        }
        return this.currentToken;
      } catch (error: any) {
        if (error.response && error.response.status === 401) {
          this.currentToken = null;
          this.currentTokenExpiry = 0;
          if (this.onInvalidSession) this.onInvalidSession();
        }
        return this.currentToken;
      } finally {
        this.tokenRequestPromise = null;
      }
    })();

    return this.tokenRequestPromise;
  }

  clearAuthToken() {
    this.currentToken = null;
    this.currentTokenExpiry = 0;
  }
}

File: auth-sdk/flutter/lib/token_manager.dart
import 'package:flutter/foundation.dart' show kIsWeb;
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'config.dart';

abstract interface class StorageAdapter {
  Future<String?> getItem(String key);
  Future<void> setItem(String key, String value);
  Future<void> removeItem(String key);
}

class DefaultSecureStorageAdapter implements StorageAdapter {
  const DefaultSecureStorageAdapter([this._storage = const FlutterSecureStorage()]);
  final FlutterSecureStorage _storage;

  @override
  Future<String?> getItem(String key) => _storage.read(key: key);

  @override
  Future<void> setItem(String key, String value) =>
      _storage.write(key: key, value: value);

  @override
  Future<void> removeItem(String key) => _storage.delete(key: key);
}

class TokenManager {
  TokenManager(this.api, this.platform)
      : storage = kIsWeb ? null : const DefaultSecureStorageAdapter();

  final AuthApi api;
  final Platform platform;
  final StorageAdapter? storage;
  String? _token;
  DateTime? _expiry;
  Future<String?>? _pending;
  final _listeners = <void Function(String?)>[];
  void Function()? onInvalidSession;

  String? get currentToken => _token;

  void onUpdate(void Function(String?) callback) => _listeners.add(callback);
  void _notify() {
    for (final listener in List.of(_listeners)) listener(_token);
  }

  void setCurrentToken(String? token) {
    _token = token;
    _expiry =
        token == null ? null : DateTime.now().add(const Duration(minutes: 15));
    _notify();
  }

  Future<String?> get() {
    if (_token != null &&
        _expiry != null &&
        DateTime.now().isBefore(_expiry!.subtract(const Duration(seconds: 10))))
      return Future.value(_token);
    return _pending ??= _fetch().whenComplete(() => _pending = null);
  }

  Future<String?> _fetch() async {
    try {
      final headers = <String, String>{'x-platform': platform.name};
      if (platform != Platform.web && storage != null) {
        final id = await storage!.getItem('sessionId');
        if (id != null) headers['x-session-id'] = id;
      }
      final data = await api.send('POST', '/token', headers: headers);
      if (platform != Platform.web &&
          storage != null &&
          data['session_id'] != null)
        await storage!.setItem('sessionId', '${data['session_id']}');
      setCurrentToken(data['token'] as String?);
      return _token;
    } on ApiException catch (error) {
      if (error.statusCode == 401) {
        setCurrentToken(null);
        onInvalidSession?.call();
      }
      return _token;
    }
  }
}
